zson/builder.go

```diff
@@ -0,0 +1,241 @@
+package zson
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+)
+
+// A Builder transforms a fully-typed Value produced by an Analyzer
+// into a zng.Value.
+type Builder struct {
+	*zcode.Builder
+}
+
+func NewBuilder() *Builder {
+	return &Builder{
+		Builder: zcode.NewBuilder(),
+	}
+}
+
+func (b *Builder) Build(val Value) (zng.Value, error) {
+	b.Reset()
+	if err := b.buildValue(val); err != nil {
+		return zng.Value{}, err
+	}
+	it := b.Bytes().Iter()
+	bytes, _, err := it.Next()
+	if err != nil {
+		return zng.Value{}, err
+	}
+	return zng.Value{val.TypeOf(), bytes}, nil
+}
+
+func (b *Builder) buildValue(val Value) error {
+	switch val := val.(type) {
+	case *Primitive:
+		return b.buildPrimitive(val)
+	case *Record:
+		return b.buildRecord(val)
+	case *Array:
+		return b.buildArray(val)
+	case *Set:
+		return b.buildSet(val)
+	case *Union:
+		return b.buildUnion(val)
+	case *Map:
+		return b.buildMap(val)
+	case *Enum:
+		return b.buildEnum(val)
+	case *TypeValue:
+		return b.buildTypeValue(val)
+	case *Null:
+		b.AppendNull()
+		return nil
+	}
+	return fmt.Errorf("unknown ast type: %T", val)
+}
+
+func (b *Builder) buildPrimitive(val *Primitive) error {
+	switch zng.AliasedType(val.Type).(type) {
+	case *zng.TypeOfUint8, *zng.TypeOfUint16, *zng.TypeOfUint32, *zng.TypeOfUint64:
+		v, err := strconv.ParseUint(val.Text, 10, 64)
+		if err != nil {
+			return fmt.Errorf("invalid unsigned integer: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeUint(v))
+		return nil
+	case *zng.TypeOfInt8, *zng.TypeOfInt16, *zng.TypeOfInt32, *zng.TypeOfInt64:
+		v, err := strconv.ParseInt(val.Text, 10, 64)
+		if err != nil {
+			return fmt.Errorf("invalid integer: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeInt(v))
+		return nil
+	case *zng.TypeOfDuration:
+		v, err := time.ParseDuration(val.Text)
+		if err != nil {
+			return fmt.Errorf("invalid duration: %s", val.Text)
+		}
+		b.AppendPrimitive(zng.EncodeDuration(int64(v)))
+		return nil
+	case *zng.TypeOfTime:
+		t, err := time.Parse(time.RFC3339Nano, val.Text)
```

Comment:
Maybe later there's an optimization where we can avoid parsing primitive values twice (I know, zng is for performance, but i think having zson be fast will be valuable).

Reply:
@henridf: Do you mean an optimization to go directly from val.Text to a zcode.Bytes value without a trip through the Go representation (like time.Time here)?

Reply:
There's a `time.Parse(..)` call in parser-values.go that's done just as a grammar test.  Then this one to actually build the `zcode.Bytes`. There's probably a few different ways to coalesce those into one, but as I noted above this can come later I think.

Reply:
Agreed.  I hadn't first worried about this because I wasn't think about this being on the critical path.  Also, the design of how constants imply their type is necessary heavy-weight.  There's probably some regexp trickery would could come up with here that would classify the type in analyzer then do the parsing work in builder.